### ISSUE TITLE
2XSC v8.1.3 - Attempting to delete data results in error when entity still in use

### ISSUE THREAD

**Issue by tpperlman** [2sic/2sxc#636]
I created some test data and a view, but can't delete the data. When I attempt to do so, I receive the following error from the API:

**URL:**

http://dnndev.me/desktopmodules/2sxc/api/eav/entities/delete?appId=2&contentType=46ad6403-63f4-4c89-a61c-da38e88837fc&id=627

**ExceptionMessage:**

"The entity 627 cannot be deleted because of it is referenced by another object."

**ExceptionType:**

"System.InvalidOperationException"

**Message:**

"An error has occurred."

StackTrace:

"   at ToSic.Eav.WebApi.EntitiesController.Delete(String contentType, Int32 id, Nullable`1 appId)

   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.<>c__DisplayClassc.<GetExecutor>b__6(Object instance, Object[] methodParameters)

   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ExecuteAsync(HttpControllerContext controllerContext, IDictionary`2 arguments, CancellationToken cancellationToken)

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Tracing.ITraceWriterExtensions.<TraceBeginEndAsyncCore>d__18`1.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Controllers.ApiControllerActionInvoker.<InvokeActionAsyncCore>d__0.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Tracing.ITraceWriterExtensions.<TraceBeginEndAsyncCore>d__18`1.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Controllers.ActionFilterResult.<ExecuteAsync>d__2.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__2.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__2.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__2.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__2.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Tracing.Tracers.HttpControllerTracer.<ExecuteAsyncCore>d__5.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Tracing.ITraceWriterExtensions.<TraceBeginEndAsyncCore>d__18`1.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Dispatcher.HttpControllerDispatcher.<SendAsync>d__1.MoveNex

**Comment by iJungleboy** [comment]
This means the entity is still in use at some point - but there is currently no nice UI to find out where it is still in use. This is just a don't-shoot-yourself error, but it could be nicer, I agree.

**Comment by iJungleboy** [comment]
So as this is actually the real error, and it's behaving as expected, it's not a bug. The only thing that would be better is an improved error message, but that's no priority from our side, so we'll close this for now.

**Comment by tpperlman** [comment]
This is still actually a bug because I still can't delete it even after I delete all the modules and clear the recycle bin. The item still exists "somewhere", but I can't see where. It is also not a demo item for the Content Type. In my example there are two data items, one is a demo and the other is the one I'd like to delete.

**Comment by tpperlman** [comment]
I actually think this causes additional support overhead since my users will be reporting the issue to me and I'll have to address it.

**Comment by iJungleboy** [comment]
Ok. Re-opened issue #610 which is about providing a better error message.

**Comment by tpperlman** [comment]
Thanks! I am also getting a 500 error when I even try to edit the data item. I assumes that the linked image from ADAM can't be found.

**URL:**

http://dnndev.me/desktopmodules/2sxc/api/dnn/Hyperlink/ResolveHyperlink?hyperlink=File%3A234

**ExceptionMessage:**
	
	"Error getting List of St... Name: ModuleDataSource"
ExceptionType
	
	"System.Exception"
InnerException
	
	Object { Message="An error has occurred.",  ExceptionMessage="Object reference not set... instance of an object.",  ExceptionType="System.NullReferenceException",  more...}
ExceptionMessage
	
	"Object reference not set to an instance of an object."
ExceptionType
	
	"System.NullReferenceException"
Message
	
	"An error has occurred."
StackTrace
	
	"   at ToSic.SexyContent.DataSources.ModuleDataSource.GetContent()

   at ToSic.Eav.DataSources.DataStream.get_List()"
Message
	
	"An error has occurred."
StackTrace
	
	"   at ToSic.Eav.DataSources.DataStream.get_List()

   at ToSic.SexyContent.AppAndDataHelpers..ctor(SexyContent sexy, ModuleInfo module, ViewDataSource data, App app)

   at ToSic.SexyContent.WebApi.SxcApiController.get_AppAndDataHelpers()

   at ToSic.SexyContent.WebApi.Dnn.HyperlinkController.ResolveHyperlink(String hyperlink)

   at lambda_method(Closure , Object , Object[] )

   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.<>c__DisplayClass10.<GetExecutor>b__9(Object instance, Object[] methodParameters)

   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ExecuteAsync(HttpControllerContext controllerContext, IDictionary`2 arguments, CancellationToken cancellationToken)

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Tracing.ITraceWriterExtensions.<TraceBeginEndAsyncCore>d__18`1.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Controllers.ApiControllerActionInvoker.<InvokeActionAsyncCore>d__0.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Tracing.ITraceWriterExtensions.<TraceBeginEndAsyncCore>d__18`1.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Controllers.ActionFilterResult.<ExecuteAsync>d__2.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__2.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__2.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__2.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__2.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Tracing.Tracers.HttpControllerTracer.<ExecuteAsyncCore>d__5.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Tracing.ITraceWriterExtensions.<TraceBeginEndAsyncCore>d__18`1.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

   at System.Web.Http.Dispatcher.HttpControllerDispatcher.<SendAsync>d__1.MoveNext()"

**Comment by iJungleboy** [comment]
Note: I accidentally re-opened this issue, but this can stay closed as I have the issue for better error message when deleting #610. So i'll re-close this.

**Comment by iJungleboy** [comment]
Regarding your missing file, this sounds like correct behavior: The UI is expecting a file, it's not there, systems says error: not there. This sounds like the correct expected behaviour. Am I missing something?

**Comment by tpperlman** [comment]
Thanks again for the quick reply. I believe that this is worthy of an exception, but I also can't seem to delete the file reference when this error occurs.

It is also my opinion that if SXC is designed to make content editing dummy-proof, it should provide better exceptions than "500 Errror". When users receive this, this is what they report to us. We can't count on users to document everything they did step-by-step to get there. Short of that, we as website admins will have to try to diagnose the issues before we can try and resolve or even report them here. I think it is vital that any content editor features should provide as much error notification as possible so that it truly can become the best content management feature in DNN.

**Comment by tpperlman** [comment]
![missing file](https://cloud.githubusercontent.com/assets/1871915/13082144/5ca1ca1a-d49d-11e5-8aa5-0a212fa29f81.JPG)

This is the error that content editors see. Isn't very descriptive until you peek at the JSON data.

**Comment by iJungleboy** [comment]
Ok, thx. Pls open a separate issue for missing-file error handling.
